### PR TITLE
[stdlib] stronger div, mod lemmas via module type

### LINKED ISSUE #16186: [stdlib] unnecessarily weak modulo lemmas in Nat
#### Description of the problem

Performing `Search ((_ * _) mod _).` after `Require Import PeanoNat.` returns the following

```coq
Nat.mod_mul:
  forall a b : nat, b <> 0 -> (a * b) mod b = 0
Nat.mul_mod_idemp_l:
  forall a b n : nat, n <> 0 -> (a mod n * b) mod n = (a * b) mod n
Nat.mul_mod_idemp_r:
  forall a b n : nat, n <> 0 -> (a * (b mod n)) mod n = (a * b) mod n
Nat.mul_mod:
  forall a b n : nat, n <> 0 -> (a * b) mod n = (a mod n * (b mod n)) mod n
Nat.mul_mod_distr_l:
  forall a b c : nat, b <> 0 -> c <> 0 -> (c * a) mod (c * b) = c * (a mod b)
Nat.mul_mod_distr_r:
  forall a b c : nat, b <> 0 -> c <> 0 -> (a * c) mod (b * c) = a mod b * c
```

Interestingly, **all** of the above `_ <> 0` conditions are unnecessary.
This is shown by the following proofs.

```coq
Lemma mod_mul' a b : (a * b) mod b = 0.
Proof. now destruct b; [rewrite Nat.mul_0_r|apply Nat.mod_mul]. Qed.

Lemma mul_mod_idemp_l' a b n : (a mod n * b) mod n = (a * b) mod n.
Proof. now destruct n; [|apply Nat.mul_mod_idemp_l]. Qed.

Lemma mul_mod_idemp_r' a b n : (a * (b mod n)) mod n = (a * b) mod n.
Proof. now destruct n; [|apply Nat.mul_mod_idemp_r]. Qed.

Lemma mul_mod' a b n : (a * b) mod n = (a mod n * (b mod n)) mod n.
Proof. now destruct n; [|apply Nat.mul_mod]. Qed.

Lemma mul_mod_distr_l' a b c : (c * a) mod (c * b) = c * (a mod b).
Proof.
  now destruct b; destruct c; [|rewrite Nat.mul_0_r| |apply Nat.mul_mod_distr_l].
Qed.

Lemma mul_mod_distr_r' a b c : (a * c) mod (b * c) = a mod b * c.
Proof.
  now destruct b; destruct c; [repeat rewrite Nat.mul_0_r..|apply Nat.mul_mod_distr_r].
Qed.
```

Currently, sequential `rewrite` using the weak lemmas creates inconvenient subgoals.
It would be nice to have stronger statements in the standard library.

At times, `div` is even worse off https://github.com/coq/coq/issues/16186#issuecomment-1158576972

#### Coq Version
`The Coq Proof Assistant, version 8.15.2`

### PR DESCRIPTION
**Pull request by mrhaandi**
This is a follow-up to uniformization of `div` and `mod` in https://github.com/coq/coq/pull/14086.
As a consequence, we can uniformly derive stronger lemmas for `Nat` and `N` assuming `a / 0 == 0` and `a mod 0 == a`.

Closes https://github.com/coq/coq/issues/16186

The basic idea is to have an abstract, optional specification
```coq
Module Type NZDivSpec0 (Import A : Eq')(Import B : ZeroSuccPred' A)(Import C : DivMod' A).
  Axiom div_0_r : forall a, a / 0 == 0.
  Axiom mod_0_r : forall a, a mod 0 == a.
End NZDivSpec0.
```
In addition, modules `NDiv0.NDivProp0` (as apposed to `NDiv.NDivProp`) and `NLcm0.NLcmProp0` (as apposed to `NLcm.NLcmProp`) provide stronger lemmas based on `NZDivSpec0`. For example: 
```coq
(* NDivProp  *) Lemma div_div : forall a b c, b~=0 -> c~=0 -> (a/b)/c == a/(b*c).
(* NDivProp0 *) Lemma Div0.div_div : forall a b c, (a/b)/c == a/(b*c).
```
Any number system which satisfies `NZDivSpec0` may for free benefit from the stronger lemmas.

- We instantiate `PeanoNat.Nat` and `BinNat.N` as above.
- We do **not** deprecate `NDivProp` or `NLcmProp`, which are valid (and useful) for number systems that do not satisfy `NZDivSpec0`.
- We do **not** touch `Z`, which is a separate concern.

### Deprecation Phases
We aim for maximal compatibility for each phase with deprecation warnings.
#### Phase 1
- Stronger lemmas are encapsulated in `Div0` and `Lcm0` modules.
- Weaker lemmas encapsulated in `Notation` with a `deprecated` attribute.
- `Search` does not display weaker lemmas due to `Private_` prefix.
#### Phase 2
- Stronger lemmas are both top-level and in `Div0` and `Lcm0` modules.
- Weaker lemmas are removed.
- `Div0` and `Lcm0` modules are deprecated.
#### Phase 3
- Modules `Div0` and `Lcm0` are removed.

  

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by mrhaandi** [comment]
For `PeanoNat` this approach seems to work really well. It provides the strongest `Nat.xyz_full` lemmas and uses only a single `Include NDivProp0.`.

**Comment by olaure01** [comment]
Could it be possible to rely on the module system to keep the same name for strong statements as for weak ones? thus avoiding the use of `_full`.
Under a given name, you import `NZDiv` you get the weak statement, you import `NZDiv0` you get the strong one.

**Comment by mrhaandi** [comment]
> Could it be possible to rely on the module system to keep the same name for strong statements as for weak ones? thus avoiding the use of `_full`. Under a given name, you import `NZDiv` you get the weak statement, you import `NZDiv0` you get the strong one.

I'm not sure what is the best way here and I am open to suggestions. Shadowing the existing `Nat` module by a different one upon an extra import is possible, but I rather prefer not to. This would lead to a lot of confusion.

A good compromise is still to have the weak lemmas marked as deprecated (as in the other PR) and after a transition rename the strong ones to their proper names.
At an early point one can hide weak lemmas from `Search`, so the user is not overwhelmed by duplicates.

**Comment by olaure01** [comment]
> I'm not sure what is the best way here and I am open to suggestions. Shadowing the existing Nat module by a different one upon an extra import is possible, but I rather prefer not to. This would lead to a lot of confusion.

Talking about the final target (not considering the deprecation phase yet), I would imagine for example creating a `NExtraProp0` in `Numbers.Natural.Abstract.NProperties` which uses `NDivProp0` instead of `NDivProp`, and then to include `NExtraProp0` instead of `NExtraProp0` in `PeanoNat.Nat`. `NDivProp0` could redefine weak statements from `NDivProp` into strong ones as it is the case currently for example with `NDiv.mod_same : forall a, a~=0 -> a mod a == 0.` overloading `NZDiv.mod_same : forall a, 0<a -> a mod a == 0.`.

**Comment by mrhaandi** [comment]
> Talking about the final target (not considering the deprecation phase yet)

Oh yes, replacing weak lemmas in `PeanoNat.Nat` by strong ones using the module system is easy; almost everything is already in place. So the final target is not a problem. Next week I will look deeper into the `Z` counterpart.

**Comment by mrhaandi** [comment]
I now added stronger `_full` lemmas via inclusion of `ZDivFloor0.v` to the `Z` module. This is cleaner than https://github.com/coq/coq/pull/16189.

@olaure01 What is the status of `ZArith/Zdiv.v`? It contains all kinds of (sometimes incoherently named) lemmas, sometimes just restating facts from `Z`. Some of the lemmas could have (without this PR) already been deprecated in favor of `Z` lemmas.

I see two options.
1) Deprecate as many `ZArith/Zdiv.v` lemmas as possible now in favor of `_full` lemmas.
2) Do not touch `ZArith/Zdiv.v`. Only deprecate weak lemmas in `Z`. After this first deprecation phase, rename strong lemmas (removing `_full` suffix). Then deprecate as many `ZArith/Zdiv.v` using strong lemmas (without the `_full` suffix).

Option 2 would be easier on the users. Option 1 would entail e.g. the `Zdiv.Zdiv_0_l` => `Z.div_0_l_full` => `Z.div_0_l` dance.

**Comment by olaure01** [comment]
> What is the status of `ZArith/Zdiv.v`?

I am not sure. Similar files like `ZArith/Zmin.v` are explicitly mentioned as deprecated/obsolete (thus probably subject to the analogue of #14736).

> 1. Deprecate as many `ZArith/Zdiv.v` lemmas as possible now in favor of `_full` lemmas.

Since deprecation phases are long (in particular when we want to reuse old names), I am in general in favor of starting them as soon as possible.

> 2. Do not touch `ZArith/Zdiv.v`. Only deprecate weak lemmas in `Z`. After this first deprecation phase, rename strong lemmas (removing `_full` suffix). Then deprecate as many `ZArith/Zdiv.v` using strong lemmas (without the `_full` suffix).

I agree that in the present situation going back and forth through `_full` names would be painful, and option 2 looks better.

Would it be reasonable to consider settling now the architecture of the final target of the process (with strong statements available under `_full`-free names) and declaring temporary modules `PeanoNatFull.Nat` and `BinIntFull.Z` to make this architecture accessible immediately? This could allow users interested in adapting now their development towards the target situation to:
 1. import now these new modules to directly use final names for strong statements;
 2. in the next step, reach the final situation by just replacing `PeanoNatFull` with `PeanoNat` and `BinIntFull` with `BinInt`.

**Comment by mrhaandi** [comment]
> `PeanoNatFull.Nat` and `BinIntFull.Z`

I'm not sure whether I like the idea. It creates a third competing module, which will surely take over a year to again deprecate.
A newcomer will have the (sometimes uninformed) choice between
1) `ZDiv` (currently often used, mostly deprecated)
2) `BinInt.Z` (partially deprecated)
3) `BinIntFull.Z` (handy, but will be deprecated shortly)

**Comment by mrhaandi** [comment]
> to be used only by project maintainers willing to prepare for future

I am such a developer of a (undecidability) library with over 100K LOC. `PeanoNatFull.Nat` would shadow `PeanoNat.Nat`, which makes imports order sensitive (also via some transitive exports). So I would not use `PeanoNatFull.Nat` for reasons of error locality.

**Comment by mrhaandi** [comment]
While trying to sideways refactor `NDivProp` I noticed that the whole module tower in `PeanoNat` is then starting to fall over. Mainly because further up module types such as `NLcmProp` contain weak lemmas e.g. `Lemma mod_divide : forall a b, b~=0 -> (a mod b == 0 <-> (b|a)).` with a dependence on the weak division lemmas. The main consequence is that `NLcmProp` and others will need a `NLcmProp0` counterpart. Otherwise, we cannot use `Notation` to deprecate weak `nat` lemmas.

Also, we cannot deprecate the module types such as `NDivProp` or `NLcmProp` but only their use in `PeanoNat`.
Otherwise, the abstract module types will become incompatible with other division by zero conventions.

I will shift my focus solely on `nat` and `N` now. `Z` has an even larger module type tower and will require further investigation.

**Comment by mrhaandi** [comment]
@olaure01 I finally figured out (for `Nat`/`N`) a module structure which has almost no code duplication and allows for deprecation via `Notation`. The overall approach is now solid for `Nat`/`N` and quire useful. I consider this close to done and will not touch `Z` in this PR. If ci agrees, I will switch the status to "Ready for review".

**Comment by Alizter** [comment]
@mrhaandi I would suspect that some boolean equality is not found and the boolean equality scheme tries to go and derive it automatically. Are you able to observe slow down in that file when you run it manually in an editor?

cc @herbelin

**Comment by mrhaandi** [comment]
> Are you able to observe slow down in that file when you run it manually in an editor?

In the editor I see no slowdown in `success/SchemeEquality.v` and the overall `test-suite` compilation time is basically the same.
To be sure, I use a clean checkout (to have no caches present). Although I doubt it, but maybe my OCaml `4.12.0` behaves differently than `4.09.0` from the ci?

**Comment by mrhaandi** [comment]
@Alizter What did you change in the ci that `success/SchemeEquality.v` passed? Or was it just a rerun?

**Comment by Alizter** [comment]
@mrhaandi I rerun it yesterday. Sometimes the CI takes its time. It was unrelated to SchemeEquality.v.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@olaure01 Did you want to have a look?

**Comment by olaure01** [comment]
@Alizter I planned to have look next week or the week after, but if it is OK with you, please go.

**Comment by Alizter** [comment]
@olaure01 In that case I will wait for your review. No rush.

File: theories/Numbers/Natural/Abstract/NGcd.v
Comment (olaure01) [reply theories/Numbers/Natural/Abstract/NGcd.v:60]: Could be `N.xxx` when using `BinNat`.

File: theories/Numbers/Natural/Abstract/NGcd.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NGcd.v:58]: This induces a type change: would it require a deprecation phase?
At least a mention in changelog I think.

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:125]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.mod_divide instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:127]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.divide_div_mul_exact instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:129]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.gcd_div_factor instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:131]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.gcd_mod instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:133]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.gcd_mod instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:135]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.lcm_equiv2 instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:137]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.divide_div instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:139]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.lcm_unique instead.")]
```

File: theories/Numbers/Natural/Abstract/NLcm0.v
Comment (olaure01) [inline theories/Numbers/Natural/Abstract/NLcm0.v:141]:
```suggestion
#[deprecated(since="8.17",note="Use Lcm0.lcm_unique_alt instead.")]
```

**Review by olaure01 — CHANGES_REQUESTED** [review]
I think the work on induction in `NZOrder` and on Bezout in `NGcd` should be documented in changelog.
You could also mention more about the future in changelog: planned deprecation phases, and the trick to directly go close to the end following: `Module Nat := Nat.Div0.`

File: theories/Numbers/NatInt/NZOrder.v
Comment (olaure01) [inline theories/Numbers/NatInt/NZOrder.v:394]:
```suggestion
intros x1 x2 H1 x3 x4 H2; unfold Rlt. now rewrite H1, H2.
```

File: theories/Numbers/NatInt/NZOrder.v
Comment (olaure01) [inline theories/Numbers/NatInt/NZOrder.v:399]:
```suggestion
intros x1 x2 H1 x3 x4 H2; unfold Rgt; now rewrite H1, H2.
```

File: theories/Numbers/Natural/Abstract/NGcd.v
Comment (mrhaandi) [reply theories/Numbers/Natural/Abstract/NGcd.v]: Yes.

File: theories/Numbers/Natural/Abstract/NGcd.v
Comment (mrhaandi) [reply theories/Numbers/Natural/Abstract/NGcd.v:58]: I added this fact to the changelog.

File: theories/Numbers/NatInt/NZOrder.v
Comment (mrhaandi) [reply theories/Numbers/NatInt/NZOrder.v]: Done.

File: theories/Numbers/NatInt/NZOrder.v
Comment (mrhaandi) [reply theories/Numbers/NatInt/NZOrder.v]: Done.

**Comment by mrhaandi** [comment]
> I think the work on induction in `NZOrder` and on Bezout in `NGcd` should be documented in changelog.

Yes, I forgot to add a note on `measure_induction` which is a uniform way to have `Wf_nat.induction_ltof1` and is compatible with the `induction` tactic. I added the info to the changelog, thank you.

> You could also mention more about the future in changelog: planned deprecation phases, and the trick to directly go close to the end following: `Module Nat := Nat.Div0.`

Yes.

**Comment by olaure01** [comment]
@coq/doc-maintainers This PR both adds, removes, changes and deprecates things. Is the changelog OK, or would it be better to split it into separated **Added** / **Changed** / **Deprecated** / ... entries?

File: doc/changelog/11-standard-library/16203-concise_mod_type.rst
Comment (Alizter) [inline doc/changelog/11-standard-library/16203-concise_mod_type.rst:13]:
```suggestion
  :g:`divide_gcd_iff`, :g:`gcd_mul_diag_l` is removed
```

**Comment by Alizter** [comment]
I think splitting up the changelog entry would be a good idea.

File: doc/changelog/11-standard-library/16203-concise_mod_type.rst
Comment (mrhaandi) [reply doc/changelog/11-standard-library/16203-concise_mod_type.rst]: Yes.

**Comment by mrhaandi** [comment]
> I think splitting up the changelog entry would be a good idea.

I now split the changelog entries into three orthogonal bullet points. Is this the standard approach, or should I create distinct changelog files?

**Comment by proux01** [comment]
That's fine, no need for distinct files.

**Review by olaure01 — APPROVED** [review]

**Comment by Alizter** [comment]
I will merge this soon. @proux01 are you still having a look?

**Comment by mrhaandi** [comment]
> I will merge this soon.

Squashed and rebased to be on the safe side.

**Comment by proux01** [comment]
> I will merge this soon. @proux01 are you still having a look?

@Alizter nope, feel free to merge

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by JasonGross** [comment]
@mrhaandi Has any investigation been done into removing the useless hypotheses from the Z lemmas?

**Comment by mrhaandi** [comment]
> Has any investigation been done into removing the useless hypotheses from the Z lemmas?

Unfortunately, the situation with `Z` is more complicated than with `nat`. Also, the current solution for `nat` has drawbacks regarding usability, in order to be backwards compatible. Since for my work `nat` is most important, I have little motivation to put comparable (or even more) effort into `Z`.

**Comment by JasonGross** [comment]
@mrhaandi would you be willing to describe what the drawbacks and considerations are?  I've been considering doing this work myself, if I ever find the time for it, and would be interested in not retreading known dead-ends and having a better sense up front what the rough design trade-offs are.

**Pull request merged**
